devito/types/object.py

```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
     @property
     def _mem_internal_lazy(self):
         return self._liveness == 'lazy'
+
+
+class PetscObject(AbstractObjectWithShape):
+
+    is_PetscObject = True
+
+    def __init__(self, name, dtype, **kwargs):
+        self.name = name
+        self._dtype = dtype
+        self._is_const = kwargs.get('is_const', False)
+
+    # taken exactly from Function class but removed the "shape_global"
+    #  # and also added the option for all dimensions, shape and grid to be None
+    @classmethod
+    def __shape_setup__(cls, **kwargs):
```

Comment:
both shape_setup and indices_setup should belong to AbstractObjectWithShape

Reply:
Does this mean I should remove the 'empty' shape_setup and indices_setup that are currently inside AbstractObjectWithShape and replace them with these ones?

Reply:
Why not? 

Say I create a new subtype, let's call it `class MyNewObject(AbstractObjectWithShape)`

I want this new type to share all the key PetscObject mechanisms , don't I?